### ISSUE TITLE
Generics in the Cache interface

### ISSUE THREAD

**Issue by abdurahmanadilovic** [kezhenxu94/cache-lite#6]
https://github.com/kezhenxu94/cache-lite/blob/9b55e496ff35cb1bb2b5765774bc5a71ce03bd92/src/main/kotlin/io/github/kezhenxu94/cache/lite/impl/ExpirableCache.kt#L31

Shouldn't the lastFlushTime be reset in the recycle function to System. nanoTime()? Otherwise the cache is expired all the time

**Comment by kezhenxu94** [comment]
@abdurahmanadilovic yes good catch, could you please open a pull request to fix this?

**Comment by abdurahmanadilovic** [comment]
@kezhenxu94 absolutely, I'll make a pull request. The only reason I noticed this issue is that I tried to incorporate this cache into my own app and I noticed that the cache will be in the expired state forever :).
Btw I liked your usage of the decorator pattern, its so simple, yet so powerful!

**Comment by kezhenxu94** [comment]
> @kezhenxu94 absolutely, I'll make a pull request. The only reason I noticed this issue is that I tried to incorporate this cache into my own app and I noticed that the cache will be in the expired state forever :).
> Btw I liked your usage of the decorator pattern, its so simple, yet so powerful!

Glad that it helps, if it's a blocker for you, I can release a new version after your patch, thanks in advance :)

**Comment by kezhenxu94** [comment]
@abdurahmanadilovic and I notice your title "Generics in the Cache interface", are you proposing to make the `Cache` interface a generic type?

**Comment by abdurahmanadilovic** [comment]
@kezhenxu94 I don't have permissions to push a new branch to this repository, is it closed by default?

Sorry about the issue title, I wanted to make another issue and propose the cache interface to use generics, something like this:
```kotlin
interface Cache<K, V> {
    val size: Int

    operator fun set(key: K, value: V)

    operator fun get(key: K): V?

    fun remove(key: K): V?

    fun clear()
}
```

But I decided agains it since its more difficult for beginners to understand. But if you think it's cool, we can add a new entry in the readme file with an example that uses generics?

**Comment by kezhenxu94** [comment]
> @kezhenxu94 I don't have permissions to push a new branch to this repository, is it closed by default?

GitHub have no such function to enable others to push a new branch to a repo which they are not collaborator, if you're interested in, I can invite you to collaborate on this project

> 
> Sorry about the issue title, I wanted to make another issue and propose the cache interface to use generics, something like this:
> 
> ```kotlin
> interface Cache<K, V> {
>     val size: Int
> 
>     operator fun set(key: K, value: V)
> 
>     operator fun get(key: K): V?
> 
>     fun remove(key: K): V?
> 
>     fun clear()
> }
> ```
> 
> But I decided agains it since its more difficult for beginners to understand. 

Yes I had thought about generifying the interface at the very implementation, but I just gave up as what you mentioned here,

> But if you think it's cool, we can add a new entry in the readme file with an example that uses generics?

Yeah, of course, we can even add another implementation (say `GenericTypedCache` or something similar) that just uses generic types, and make the original `Cache implements GenericTypedCache<Any, Any>`.

Feel free to send a pull request and we can discuss there :)

**Comment by abdurahmanadilovic** [comment]
Thanks for the merge.

> Yeah, of course, we can even add another implementation (say GenericTypedCache or something similar) that just uses generic types, and make the original Cache implements GenericTypedCache<Any, Any>.

I'll make another pull request then with a new generic GenericyTypedCache interface. I guess we can keep this issue open for a little while?

**Comment by kezhenxu94** [comment]
> I'll make another pull request then with a new generic GenericyTypedCache interface.

Cool, look forward to your work

> I guess we can keep this issue open for a little while?

Sure, let's just keep it open

**Comment by kezhenxu94** [comment]
fixed by #9
